eth/tracers/firehose.go

```diff
@@ -280,12 +291,46 @@ func NewFirehose(config *FirehoseConfig) *Firehose {
 		}
 	}
 
-	if config.ConcurrentBlockFlushing {
-		log.Info("Firehose concurrent block flushing enabled, starting block " +
-			"print worker goroutine")
-		firehose.blockPrintQueue = make(chan *blockPrintJob, 100)
-		firehose.blockFlushDone.Add(1)
-		go firehose.blockPrintWorker()
+	if config.ConcurrentBlockFlushing > 0 {
+		log.Info(fmt.Sprintf("Firehose concurrent block flushing enabled, starting %d worker goroutine", config.ConcurrentBlockFlushing))
+
+		firehose.flushJobQueue = make(chan *blockPrintJob, firehose.flushBufferSize)
+		firehose.flushOrderedOutputQueue = make(chan *blockOutput, firehose.flushBufferSize)
+		for i := 0; i < config.ConcurrentBlockFlushing; i++ {
+			firehose.flushJobWG.Add(1)
+			go firehose.blockPrintWorker()
+		}
+
+		// Output channel to order the flushing linearly
+		firehose.flushOrderedOutputWG.Add(1)
+		go func() {
+			defer firehose.flushOrderedOutputWG.Done()
+
+			buffer := make(map[uint64][]byte)
+
+			firehose.flushStartCond.L.Lock()
+			for !firehose.flushStarted {
+				firehose.flushStartCond.Wait()
+			}
```

Comment:
Couldn't we wait a message from a channel here, rarely saw a for loop that then call a `Wait`. We actually almost never used `sync.Cond` in the past AFAIK.